### PR TITLE
W9-D1-DinhTruong-Unit test

### PR DESCRIPTION
**Pull request by at-dinhtruong**
#55

### CODE REVIEW

File: app/src/at-dinhtruong/res/layout/activity_unit_test.xml [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            android:layout_marginLeft="20dp"
+            android:layout_marginRight="20dp"
+            android:background="@drawable/bg_incorrect_password"
+            android:padding="10dp"
+            android:text="@string/incorrectPassword"
+            android:textAlignment="center"
+            android:textColor="@color/colorRed"
+            android:textSize="20sp"
```
Comment (at-trungnguyen) [inline app/src/at-dinhtruong/res/layout/activity_unit_test.xml]: convention is `dp`?

File: app/src/at-dinhtruong/java/asiantech/internship/summer/unittest/UnitTestActivity.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        mEdtUsername = findViewById(R.id.editUserName);
+        mEdtPassword = findViewById(R.id.edtPassword);
+        mTvNotificationError = findViewById(R.id.tvNotificationError);
+        mBtnLogin = findViewById(R.id.btnLogin);
+    }
+
+    @Override
+    public void onClick(View view) {
+        mTvNotificationError.setVisibility(view.getVisibility());
+        String username = mEdtUsername.getText().toString();
```
Comment (at-trungnguyen) [inline app/src/at-dinhtruong/java/asiantech/internship/summer/unittest/UnitTestActivity.java]: ` mEdtUsername.getText().toString().trim();`

File: app/src/at-dinhtruong/java/asiantech/internship/summer/unittest/UnitTestActivity.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        String password = mEdtPassword.getText().toString();
```
Comment (at-trungnguyen) [inline app/src/at-dinhtruong/java/asiantech/internship/summer/unittest/UnitTestActivity.java]: `mEdtPassword.getText().toString().trim();`

File: app/src/at-dinhtruong/java/asiantech/internship/summer/unittest/UnitTestActivity.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        User user = new User(username, password);
+        switch (view.getId()) {
```
Comment (at-trungnguyen) [inline app/src/at-dinhtruong/java/asiantech/internship/summer/unittest/UnitTestActivity.java]: use `if ()`

File: app/src/at-dinhtruong/java/asiantech/internship/summer/unittest/Validate.java [hunk 5]
```diff
@@ -0,0 +1,178 @@
+package asiantech.internship.summer.unittest;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import asiantech.internship.summer.R;
+import asiantech.internship.summer.models.User;
+
+class Validate {
```
Comment (at-trungnguyen) [inline app/src/at-dinhtruong/java/asiantech/internship/summer/unittest/Validate.java]: `public final class Validate {`

File: app/src/at-dinhtruong/java/asiantech/internship/summer/unittest/Validate.java [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+    static final String USERNAME_9_20_CHARATER = "Username should have 9 to 20 characters";
```
Comment (at-trungnguyen) [inline app/src/at-dinhtruong/java/asiantech/internship/summer/unittest/Validate.java]: `public static final`

File: app/src/at-dinhtruong/java/asiantech/internship/summer/unittest/Validate.java [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+    static final String USERNAME_SHOULD_NOT_HAVE_SPECIAL_CHARACTER = "Username should not have special characters";
+    static final String USERNAME_SHOULD_HAVE_AT_MOST_2_NUMBER = "Username should have at most 2 number";
+    static final String USERNAME_SHOULD_NOT_START_WHITH_UPPER_CASE = "Username do not start with uppercase or numbers";
+    static final String USERNAME_AND_PASSWORD_ARE_DIFFERENT = "Username and password are different";
+    static final String PASSWORD_SHOULD_HAVE_2_SPECIAL_OR_NUMBER = "Password should have at least 2 special characters or numbers";
+    static final String PASSWORD_SHOULD_NOT_LOOP_CHARACTER_TWO_TIMES = "Password should not loop character more than two times";
+    static final String PASSWORD_SHOULD_NOT_END_DIGIT_OR_SEPECIAL_CHAR = "Password do dot end with digit or special character";
+    static final String PASSWORD_AT_LEAST_3_UPPERCASE_NOT_CONSECUTIVE = "Password at least three uppercase not consecutive";
+
+    static boolean isUsernameLength(String username) {
```
Comment (at-trungnguyen) [inline app/src/at-dinhtruong/java/asiantech/internship/summer/unittest/Validate.java]: `public static boolean`

Comment (at-trungnguyen) [reply app/src/at-dinhtruong/java/asiantech/internship/summer/unittest/Validate.java]: check all method bellow.

File: app/src/at-dinhtruong/java/asiantech/internship/summer/unittest/Validate.java (hunk 5 above)
Comment (at-trungnguyen) [reply app/src/at-dinhtruong/java/asiantech/internship/summer/unittest/Validate.java]: add private constructor
```
private Validate() {
    // no-op
}
```

File: app/src/at-dinhtruong/java/asiantech/internship/summer/unittest/UnitTestActivity.java [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+    private void initViews() {
+        mEdtUsername = findViewById(R.id.editUserName);
+        mEdtPassword = findViewById(R.id.edtPassword);
+        mTvNotificationError = findViewById(R.id.tvNotificationError);
+        mBtnLogin = findViewById(R.id.btnLogin);
+    }
+
+    @Override
+    public void onClick(View view) {
+        mTvNotificationError.setVisibility(view.getVisibility());
```
Comment (at-trungnguyen) [inline app/src/at-dinhtruong/java/asiantech/internship/summer/unittest/UnitTestActivity.java]: why do you need this `mTvNotificationError.setVisibility(view.getVisibility());`?

File: app/src/at-dinhtruong/res/drawable/bg_form_login.xml [hunk 9]
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="utf-8"?>
+<shape xmlns:android="http://schemas.android.com/apk/res/android">
```
Comment (at-trungnguyen) [inline app/src/at-dinhtruong/res/drawable/bg_form_login.xml]: format this file

File: app/src/at-dinhtruong/res/layout/activity_unit_test.xml (hunk 1 above)
Comment (at-trungnguyen) [reply app/src/at-dinhtruong/res/layout/activity_unit_test.xml]: put in `dimens.xml` file

File: app/src/at-dinhtruong/res/layout/activity_unit_test.xml [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+        <EditText
+            android:id="@+id/editUserName"
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            android:layout_margin="20dp"
+            android:autofillHints="@string/userName"
+            android:background="@drawable/bg_username_password"
+            android:inputType="textNoSuggestions"
+            android:padding="10dp"
+            tools:ignore="LabelFor,UnusedAttribute" />
```
Comment (at-trungnguyen) [inline app/src/at-dinhtruong/res/layout/activity_unit_test.xml]: `tools:ignore` avoid use `tools`, need fix warning.

File: app/src/at-dinhtruong/res/layout/activity_unit_test.xml [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+
+        <EditText
+            android:id="@+id/edtPassword"
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            android:layout_marginLeft="20dp"
+            android:layout_marginRight="20dp"
+            android:autofillHints="@string/userName"
+            android:background="@drawable/bg_username_password"
+            android:inputType="textNoSuggestions"
```
Comment (at-trungnguyen) [inline app/src/at-dinhtruong/res/layout/activity_unit_test.xml]: should be `android:inputType="textPassword"`

**Comment by at-trungnguyen** [comment]
@at-dinhtruong Please check comments and fix conflict.

File: app/src/at-dinhtruong/java/asiantech/internship/summer/unittest/UnitTestActivity.java (hunk 8 above)
Comment (at-dinhtruong) [reply app/src/at-dinhtruong/java/asiantech/internship/summer/unittest/UnitTestActivity.java]: First I want to keep textview error hidden, then if there are errors, the textview is visible.

**Comment by at-trungnguyen** [comment]
@at-dinhtruong LGTM!

**Pull request merged**